src/Aspire.Hosting/Orchestrator/ApplicationOrchestrator.cs

```diff
@@ -265,10 +266,19 @@ private async Task ProcessResourceUrlCallbacks(IResource resource, CancellationT
 
     private async Task OnResourceEndpointsAllocated(ResourceEndpointsAllocatedEvent @event, CancellationToken cancellationToken)
     {
-        await ProcessResourceWithoutLifetime(@event.Resource, cancellationToken).ConfigureAwait(false);
         await PublishResourceEndpointUrls(@event.Resource, cancellationToken).ConfigureAwait(false);
     }
 
+    private async Task OnResourceInitialized(InitializeResourceEvent @event, CancellationToken cancellationToken)
+    {
+        if (@event.Resource is not IResourceWithoutLifetime)
+        {
+            return;
+        }
+
+        await ProcessResourceWithoutLifetime(@event.Resource, cancellationToken).ConfigureAwait(false);
```

Comment:
We may have to defer resources that reference endpoints to be resolved later:

https://github.com/dotnet/aspire/blob/5f415f762891ee253f42a5940c0cce948ac3b8fd/playground/ParameterEndToEnd/ParameterEndToEnd.AppHost/Program.cs#L24